packages/next/build/swc/src/hook_optimizer.rs

```diff
@@ -0,0 +1,127 @@
+use regex::Regex;
+use swc_common::DUMMY_SP;
+use swc_ecmascript::ast::{
+  ArrayPat, Decl, Expr, ExprOrSuper, Ident, ImportDecl, ImportSpecifier, KeyValuePatProp, Number,
+  ObjectPat, ObjectPatProp, Pat, PropName, VarDecl, VarDeclarator,
+};
+use swc_ecmascript::visit::{Fold, FoldWith};
+
+pub fn hook_optimizer() -> impl Fold {
+  HookOptimizer::default()
+}
+
+#[derive(Debug, Default)]
+struct HookOptimizer {
+  hooks: Vec<String>,
+}
+
+impl Fold for HookOptimizer {
+  // Find hooks imported from react/preact
+  fn fold_import_decl(&mut self, decl: ImportDecl) -> ImportDecl {
+    let hook_re = Regex::new(r"^use[A-Z]").unwrap();
+    let ImportDecl {
+      ref src,
+      ref specifiers,
+      ..
+    } = decl;
+    if &src.value == "react" || &src.value == "preact/hooks" {
+      for specifier in specifiers {
+        if let ImportSpecifier::Named(named_specifier) = specifier {
+          if hook_re.is_match(&named_specifier.local.sym) {
+            self
+              .hooks
+              .push(String::from(&named_specifier.local.sym as &str))
+          }
+        }
+      }
+    }
+
+    decl
+  }
+  // Transform array desctructing to object destructuring for relevant hooks
+  fn fold_decl(&mut self, node: Decl) -> Decl {
+    let node = node.fold_children_with(self);
+    match node {
+      Decl::Var(VarDecl {
+        decls,
+        span,
+        kind,
+        declare,
+      }) => {
+        let mut new_decls: Vec<VarDeclarator> = Vec::new();
+        for decl in decls {
+          new_decls.push(self.get_decl(decl));
+        }
+
+        Decl::Var(VarDecl {
+          decls: new_decls,
+          span,
+          kind,
+          declare,
+        })
+      }
+      _ => node,
+    }
+  }
+}
+
+impl HookOptimizer {
+  fn get_decl(&mut self, decl: VarDeclarator) -> VarDeclarator {
+    let VarDeclarator {
+      name,
+      init,
+      span,
+      definite,
+    } = decl.clone();
+    if let Pat::Array(a) = name {
+      // ? Is there a way to do this without cloning
+      if let Expr::Call(c) = *init.clone().unwrap() {
+        if let ExprOrSuper::Expr(i) = c.callee {
+          if let Expr::Ident(Ident { sym, .. }) = *i {
+            if self.hooks.contains(&sym.to_string()) {
```

Comment:
Using `JsWord` for `self.hooks` can be better, as comparison of `JsWord` is simple memory comparison while the comparison of string requires `memcmp`.

Reply:
I tried doing this but got in over my head with lifetimes. I will need to read up a little bit more.

Reply:
You can refer to https://rustdoc.swc.rs/swc_atoms/ to see what is `JsWord`

Reply:
Thanks, my issue was using `JsWordStaticSet` instead of `JsWord`.